internal/app/machined/pkg/controllers/runtime/oom.go

```diff
@@ -0,0 +1,208 @@
+type OOMController struct {
+	CgroupRoot      string
+	ActionTriggered time.Time
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *OOMController) Name() string {
+	return "runtime.OOMController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *OOMController) Inputs() []controller.Input {
+	return nil
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *OOMController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+func (ctrl *OOMController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	ticker := time.NewTicker(sampleInterval)
+	tickerC := ticker.C
+
+	if ctrl.CgroupRoot == "" {
+		ctrl.CgroupRoot = constants.CgroupMountPath
+	}
+
+	for {
+		// the controller runs a single time
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		case <-tickerC:
+		}
+
+		node, err := cgroups.GetCgroupProperty(constants.CgroupMountPath, mempressureProp)
+		if err != nil {
+			fmt.Println("cannot read memory pressure", err)
+			continue
+		}
+
+		fmt.Println(node.MemoryPressure)
+
+		spans, ok := node.MemoryPressure[pressureType]
+		if !ok {
+			fmt.Println("cannot find memory pressure type:", pressureType)
+			continue
+		}
+
+		value, ok := spans[pressureSpan]
+		if !ok {
+			fmt.Println("cannot find memory pressure span:", pressureSpan)
+			continue
+		}
+
+		if !value.IsSet || value.IsMax {
+			continue
+		}
+
+		val, err := strconv.ParseFloat(value.String(), 32)
+		if err != nil {
+			fmt.Println("cannot parse memory pressure:", pressureSpan, err)
+			continue
+		}
+		fmt.Println("monitoring", value.String(), val, err)
+
+		if val > psiThresh && time.Since(ctrl.ActionTriggered) > cooldownTimeout {
+			ctrl.ActionTriggered = time.Now()
+			ctrl.OomAction()
+		}
+	}
+}
+
+// OomAction
+func (ctrl *OOMController) OomAction() {
+	fmt.Println("OOM action!")
+
+	ranking := []oomRankedCgroup{}
+
+	for _, cg := range []struct {
+		dir   string
+		class int
+	}{
+		{"kubepods/besteffort", OomCgroupClassBesteffort},
+		{"kubepods/burstable", OomCgroupClassBurstable},
+		{"kubepods/guaranteed", OomCgroupClassGuaranteed},
+		{"podruntime", OomCgroupClassPodruntime},
+		{"system", OomCgroupClassSystem},
+	} {
+		entries, err := os.ReadDir(filepath.Join(constants.CgroupMountPath, cg.dir))
+		if err != nil {
+			fmt.Println("cannot list cgroup members", cg.dir, err)
+			continue
+		}
+
+		for _, leaf := range entries {
+			if !leaf.IsDir() {
+				continue
+			}
+
+			leafDir := filepath.Join(constants.CgroupMountPath, cg.dir, leaf.Name())
+
+			node := cgroups.Node{}
+
+			for _, prop := range []string{"memory.current", "memory.peak", "memory.max"} {
+				err := cgroups.ReadCgroupfsProperty(&node, leafDir, prop)
+				if err != nil {
+					fmt.Println("cannot read property for cgroup", leafDir, prop, err)
+					continue
+				}
+			}
+
+			ranking = append(ranking, oomRankedCgroup{
```

Comment:
from testability perspective, I think we should extract this type to some `internal/oom` subpackage (as a public type there), implement a `Compare` function on it (whether we use CEL or do some builtin weighting), and add unit-tests for it.

It's harder (but still possible) to test the code which gathers this info from cgroupfs, but the code which makes a decision is absolutely possible to test.

Reply:
Gathering info should also be possible with a mock fs or storing mock data in a temporary dir